### PR TITLE
DEX-492 withAuth refactoring

### PR DESCRIPTION
**Pull request by DenisSerebryansky**
* DEX's withAuth introduced
* apiKey now passed to the MatcherApiRoute/V1 as optional byte array
* Checking apiKeyHash during matcher's start
* ApiKeyIsNoProvided and ApiKeyIsNotValid matcher errors introduced

### CODE REVIEW

File: dex/src/main/resources/application.conf [hunk 1]
```diff
@@ -32,7 +32,7 @@ waves.dex {
     # Bind port
     port = 6886
 
-    # Hash of API key string
+    # Hash of API key string (base58)
```
Comment (vsuharnikov) [inline dex/src/main/resources/application.conf]: 👍

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala [hunk 2]
```diff
@@ -449,6 +449,10 @@ case object NonPositiveAssetRate extends MatcherError(rate, commonEntity, outOfB
 case class RateNotFound(theAsset: Asset)
     extends MatcherError(rate, commonEntity, notFound, e"The rate for the asset ${'assetId -> theAsset} was not specified")
 
+case object ApiKeyIsNotProvided extends MatcherError(auth, commonEntity, notProvided, e"API key is not provided")
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: Suggest "API key is not provided in the configuration, please contact with the administrator"

Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/api/http/AuthRoute.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import com.wavesplatform.dex.api.{MatcherResponse, SimpleErrorResponse}
+import com.wavesplatform.dex.error.{ApiKeyIsNotProvided, ApiKeyIsNotValid}
+import com.wavesplatform.http.{`X-Api-Key`, api_key}
+
+trait AuthRoute { this: ApiRoute =>
+
+  protected val apiKeyHash: Option[Array[Byte]]
+
+  def withAuth(implicit trm: ToResponseMarshaller[MatcherResponse]): Directive0 = {
+    apiKeyHash.fold[Directive0] { complete(SimpleErrorResponse(StatusCodes.InternalServerError, ApiKeyIsNotProvided)) } { hashFromSettings =>
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/api/http/AuthRoute.scala]: :+1:

File: dex/src/main/scala/com/wavesplatform/dex/api/http/AuthRoute.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      optionalHeaderValueByType[`X-Api-Key`](()).flatMap {
+        case Some(key) if java.util.Arrays.equals(crypto secureHash key.value, hashFromSettings) => pass
+        case _ =>
+          optionalHeaderValueByType[api_key](()).flatMap {
+            case Some(key) if java.util.Arrays.equals(crypto secureHash key.value, hashFromSettings) => pass
+            case _                                                                                   => complete { SimpleErrorResponse(StatusCodes.InternalServerError, ApiKeyIsNotValid) }
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/api/http/AuthRoute.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/Matcher.scala [hunk 5]
```diff
@@ -354,13 +354,21 @@ class Matcher(settings: MatcherSettings, gRPCExtensionClient: DEXClient)(implici
         }
     }
 
+    def checkApiKeyHash(): Future[Option[Array[Byte]]] = {
+      Option(settings.restApi.apiKeyHash).filter(_.nonEmpty) match {
+        case Some(keyStr) => Base58.tryDecode(keyStr).fold(Future.failed, keyBytes => Future.successful { Some(keyBytes) })
+        case _            => Future.successful(None)
+      }
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/Matcher.scala]: It works

**Pull request merged**